components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1776,53 +1793,8 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             }
             return Ok(true);
         }
-
-        let state_key = keys::region_state_key(region_id);
-        let state: RegionLocalState = match self.ctx.engines.kv.get_msg_cf(CF_RAFT, &state_key) {
-            Err(e) => {
-                error!(
-                    "failed to load region state, ignore";
-                    "region_id" => self.fsm.region_id(),
-                    "peer_id" => self.fsm.peer_id(),
-                    "err" => %e,
-                    "target_region_id" => region_id,
-                );
-                return Ok(false);
-            }
-            Ok(None) => {
-                info!(
-                    "seems to merge into a new replica of region, let's wait.";
-                    "region_id" => self.fsm.region_id(),
-                    "peer_id" => self.fsm.peer_id(),
-                    "target_region_id" => region_id,
-                );
-                return Ok(false);
-            }
-            Ok(Some(state)) => state,
-        };
-        if state.get_state() != PeerState::Tombstone {
-            info!(
-                "wait for region split";
-                "region_id" => self.fsm.region_id(),
-                "peer_id" => self.fsm.peer_id(),
-                "target_region_id" => region_id,
-            );
-            return Ok(false);
-        }
-
-        let tombstone_region = state.get_region();
-        if tombstone_region.get_region_epoch().get_conf_ver()
-            < target_region.get_region_epoch().get_conf_ver()
-        {
-            info!(
-                "seems to merge into a new replica of region, let's wait.";
-                "region_id" => self.fsm.region_id(),
-                "peer_id" => self.fsm.peer_id(),
-                "target_region_id" => region_id,
-            );
-            return Ok(false);
-        }
-
+        // Target peers must all exist before merging and the target peer is not in region map,
+        // it means the target peer has already destroyed.
```

Comment:
Returning error directly can result in wrong state if pd provides false guarantee. For example, leader applies merge while the rest re-elect leader and rollback merge.